File: packages/solar_flutter/test/visual/compare.dart
// Compares what a widget paints with what the oracle says Figma draws. The Flutter twin of
// packages/components/test/visual/compare.mjs, with the same tolerances.

import 'dart:convert';
import 'dart:io';

import 'package:flutter/painting.dart';

/// The oracle for one component, `spec/verify/<name>.json`, read from the repository.
Map<String, dynamic> loadOracle(String name) =>
    jsonDecode(File('../../spec/verify/$name.json').readAsStringSync())
        as Map<String, dynamic>;

/// An oracle colour, `#rrggbb[aa]` or `transparent`, as a [Color].
Color oracleColour(String value) {
  if (value == 'transparent') return const Color(0x00000000);
  final hex = value.substring(1);
  final rgb = int.parse(hex.substring(0, 6), radix: 16);
  final alpha = hex.length == 8 ? int.parse(hex.substring(6), radix: 16) : 255;
  return Color((alpha << 24) | rgb);
}

/// A colour component in 0..1 as the 8-bit step it rounds to.
int _eight(double c) => (c * 255.0).round().clamp(0, 255);

/// Two colours agree within one 8-bit step; two clear colours agree whatever their RGB.
bool sameColour(Color a, Color b) {
  if (_eight(a.a) <= 1 && _eight(b.a) <= 1) return true;
  return (_eight(a.r) - _eight(b.r)).abs() <= 1 &&
      (_eight(a.g) - _eight(b.g)).abs() <= 1 &&
      (_eight(a.b) - _eight(b.b)).abs() <= 1 &&
      (_eight(a.a) - _eight(b.a)).abs() <= 3;
}

/// A CSS box-shadow list, as the oracle spells it (`0px 1px 1px 0px rgba(0, 0, 0, 0.05)`).
List<BoxShadow> oracleShadows(String value) {
  if (value == 'none' || value.isEmpty) return const [];
  final parts = <String>[];
  var depth = 0;
  var start = 0;
  for (var i = 0; i < value.length; i++) {
    if (value[i] == '(') depth++;
    if (value[i] == ')') depth--;
    if (value[i] == ',' && depth == 0) {
      parts.add(value.substring(start, i));
      start = i + 1;
    }
  }
  parts.add(value.substring(start));
  return [
    for (final part in parts)
      () {
        final colour = RegExp(r'rgba?\(([^)]*)\)').firstMatch(part);
        final c = colour!.group(1)!.split(',').map((s) => s.trim()).toList();
        final lengths = part
            .replaceFirst(colour.group(0)!, '')
            .trim()
            .split(RegExp(r'\s+'))
            .map((t) => double.parse(t.replaceFirst('px', '')))
            .toList();
        return BoxShadow(
          color: Color.fromRGBO(
            int.parse(c[0]),
            int.parse(c[1]),
            int.parse(c[2]),
            c.length > 3 ? double.parse(c[3]) : 1,
          ),
          offset: Offset(lengths[0], lengths[1]),
          blurRadius: lengths[2],
          spreadRadius: lengths.length > 3 ? lengths[3] : 0,
        );
      }(),
  ];
}

bool _sameShadows(List<BoxShadow> a, List<BoxShadow> b) =>
    a.length == b.length &&
    [
      for (var i = 0; i < a.length; i++)
        (a[i].offset - b[i].offset).distance <= 0.5 &&
            (a[i].blurRadius - b[i].blurRadius).abs() <= 0.5 &&
            (a[i].spreadRadius - b[i].spreadRadius).abs() <= 0.5 &&
            sameColour(a[i].color, b[i].color),
    ].every((ok) => ok);

/// Whether a painted value is what the oracle expects for a property.
bool agrees(String property, Object? figma, Object? painted) {
  if (painted == null) return false;
  switch (property) {
    case 'background':
    case 'borderColor':
    case 'color':
      return sameColour(oracleColour(figma! as String), painted as Color);
    case 'shadow':
      return _sameShadows(
        oracleShadows(figma! as String),
        painted as List<BoxShadow>,
      );
    case 'fontFamily':
      // Flutter names a package's font `packages/<package>/<family>`.
      return (painted as String).split('/').last == figma;
    case 'fontWeight':
      return (painted as FontWeight).value == figma;
    case 'textDecoration':
      return (painted as String) == figma;
    case 'letterSpacing':
      return ((painted as num) - (figma! as num)).abs() <= 0.02;
    default:
      return ((painted as num) - (figma! as num)).abs() <= 0.5;
  }
}

/// The oracle properties a renderer is measured on; the rest describe composition.
const measured = [
  'background',
  'borderColor',
  'borderWidth',
  'radius',
  'shadow',
  'paddingTop',
  'paddingRight',
  'paddingBottom',
  'paddingLeft',
  'gap',
  'width',
  'height',
  'color',
  'fontFamily',
  'fontWeight',
  'fontSize',
  'lineHeight',
  'letterSpacing',
  'textDecoration',
];

/// One difference: a failure, or an excused gap when [finding] is set.
class Difference {
  Difference(
    this.variant,
    this.layer,
    this.property,
    this.figma,
    this.painted, [
    this.finding,
  ]);

  final String variant;
  final String layer;
  final String property;
  final Object? figma;
  final Object? painted;
  final String? finding;

  Map<String, Object?> toJson() => {
    'variant': variant,
    'layer': layer,
    'property': property,
    'figma': figma,
    'painted': '$painted',
    if (finding != null) 'finding': finding,
  };

  @override
  String toString() =>
      '$variant · $layer.$property: Figma $figma, painted $painted';
}

/// Every measured property of one layer. A property the oracle has and the measurement lacks is a
/// failure: a platform that cannot measure something must say so, not skip it.
void compareLayer(
  String variant,
  String layer,
  Map<String, dynamic> expected,
  Map<String, Object?> painted,
  List<dynamic> excused,
  List<Difference> failures,
  List<Difference> gaps,
) {
  for (final property in measured) {
    if (!expected.containsKey(property)) continue;
    if (property == 'borderColor' && expected['borderWidth'] == 0) continue;
    final figma = expected[property];
    final excuse = excused.cast<Map<String, dynamic>>().where(
      (e) => e['layer'] == layer && e['property'] == property,
    );
    if (excuse.isNotEmpty) {
      gaps.add(
        Difference(
          variant,
          layer,
          property,
          figma,
          painted[property],
          excuse.first['finding'] as String,
        ),
      );
      continue;
    }
    if (!agrees(property, figma, painted[property])) {
      failures.add(
        Difference(variant, layer, property, figma, painted[property]),
      );
    }
  }
}

/// Writes a report beside the build output, where CI can keep it.
void report(String name, List<Difference> differences) {
  final file = File('build/visual/$name.json')..createSync(recursive: true);
  file.writeAsStringSync(
    const JsonEncoder.withIndent('  ')
        .convert(differences.map((d) => d.toJson()).toList()),
  );
}

File: packages/components/test/visual/compare.mjs
/**
 * Compares what a browser computed with what the oracle says Figma draws. Pure functions of the
 * two, so they are unit-tested on their own (compare.test.mjs) as well as used by the visual spec.
 */

/** A CSS or oracle colour as [r, g, b, a], a in 0..1; `transparent` and `none` are clear. */
export function rgba(value) {
  const v = String(value).trim().toLowerCase();
  if (v === 'transparent' || v === 'none') return [0, 0, 0, 0];
  let m = /^#([0-9a-f]{6})([0-9a-f]{2})?$/.exec(v);
  if (m)
    return [
      ...[0, 2, 4].map((i) => parseInt(m[1].slice(i, i + 2), 16)),
      m[2] === undefined ? 1 : parseInt(m[2], 16) / 255,
    ];
  m =
    /^rgba?\(\s*([\d.]+)[,\s]+([\d.]+)[,\s]+([\d.]+)(?:\s*[,/]\s*([\d.]+%?))?\s*\)$/.exec(
      v,
    );
  if (m) {
    const a =
      m[4] === undefined
        ? 1
        : m[4].endsWith('%')
          ? Number(m[4].slice(0, -1)) / 100
          : Number(m[4]);
    return [Number(m[1]), Number(m[2]), Number(m[3]), a];
  }
  throw new Error(`not a colour: ${value}`);
}

/** Two colours agree within a step of 8-bit rounding; two clear colours agree whatever their RGB. */
export function sameColour(a, b) {
  const [x, y] = [rgba(a), rgba(b)];
  if (x[3] < 0.005 && y[3] < 0.005) return true;
  return (
    x.slice(0, 3).every((c, i) => Math.abs(c - y[i]) <= 1) &&
    Math.abs(x[3] - y[3]) <= 0.01
  );
}

/** A length in pixels; `normal` is 0 for spacing, and anything else unreadable is NaN. */
export function pixels(value) {
  if (typeof value === 'number') return value;
  const v = String(value).trim();
  if (v === 'normal' || v === '0') return 0;
  const m = /^(-?[\d.]+)px$/.exec(v);
  return m ? Number(m[1]) : NaN;
}

/** A `box-shadow` value, CSS's order or Figma's, as [{x, y, blur, spread, colour}]. */
export function shadows(value) {
  const v = String(value).trim();
  if (v === 'none' || v === '') return [];
  // Split on the commas between shadows, not the ones inside a colour function.
  const parts = [];
  let depth = 0;
  let start = 0;
  for (let i = 0; i < v.length; i++) {
    if (v[i] === '(') depth++;
    if (v[i] === ')') depth--;
    if (v[i] === ',' && depth === 0) {
      parts.push(v.slice(start, i));
      start = i + 1;
    }
  }
  parts.push(v.slice(start));
  return parts.map((part) => {
    const colour =
      /(rgba?\([^)]*\)|#[0-9a-f]{3,8})/i.exec(part)?.[1] ?? 'transparent';
    const [x = 0, y = 0, blur = 0, spread = 0] = part
      .replace(colour, '')
      .trim()
      .split(/\s+/)
      .filter((t) => /px$|^0$/.test(t))
      .map(pixels);
    return { x, y, blur, spread, colour };
  });
}

function sameShadow(a, b) {
  const [x, y] = [shadows(a), shadows(b)];
  return (
    x.length === y.length &&
    x.every(
      (s, i) =>
        ['x', 'y', 'blur', 'spread'].every(
          (k) => Math.abs(s[k] - y[i][k]) <= 0.5,
        ) && sameColour(s.colour, y[i].colour),
    )
  );
}

/** Splits a CSS list on its commas, not the ones inside a colour function. */
function listOf(v) {
  const parts = [];
  let depth = 0;
  let start = 0;
  for (let i = 0; i < v.length; i++) {
    if (v[i] === '(') depth++;
    if (v[i] === ')') depth--;
    if (v[i] === ',' && depth === 0) {
      parts.push(v.slice(start, i).trim());
      start = i + 1;
    }
  }
  parts.push(v.slice(start).trim());
  return parts;
}

/**
 * A `linear-gradient(…)`, the browser's or the oracle's, as its direction and its stops, each a
 * colour at a place in percent; null for a value that is none.
 */
export function gradient(value) {
  const m = /^linear-gradient\((.*)\)$/.exec(String(value).trim());
  if (!m) return null;
  const parts = listOf(m[1]);
  const direction = /^(to |-?[\d.]+deg$)/.test(parts[0])
    ? parts.shift()
    : 'to bottom';
  const stops = parts.map((p) => {
    const i = p.lastIndexOf(' ');
    return { colour: p.slice(0, i), at: Number.parseFloat(p.slice(i + 1)) };
  });
  return { direction, stops };
}

/** Two gradients agree: one direction, and each stop's colour and place, within rounding. */
function sameGradient(a, b) {
  const [x, y] = [gradient(a), gradient(b)];
  return Boolean(
    x &&
    y &&
    x.direction === y.direction &&
    x.stops.length === y.stops.length &&
    x.stops.every(
      (s, i) =>
        Math.abs(s.at - y.stops[i].at) <= 0.5 &&
        sameColour(s.colour, y.stops[i].colour),
    ),
  );
}

/** The first family in a CSS font-family list, unquoted. */
const firstFamily = (list) =>
  String(list)
    .split(',')[0]
    .trim()
    .replace(/^["']|["']$/g, '');

/**
 * Whether a rendered value is what the oracle expects for a property.
 * @returns {boolean}
 */
export function matches(property, figma, rendered) {
  switch (property) {
    case 'background':
      // A gradient is compared as one (Table's fade); a colour as a colour.
      if (String(figma).startsWith('linear-gradient('))
        return sameGradient(figma, rendered);
      return sameColour(figma, rendered);
    case 'borderColor':
    case 'color':
      return sameColour(figma, rendered);
    case 'shadow':
      return sameShadow(figma, rendered);
    case 'fontFamily':
      return firstFamily(rendered) === figma;
    case 'fontWeight':
      return Number(rendered) === Number(figma);
    case 'textDecoration':
      return String(rendered).split(' ')[0] === figma;
    case 'letterSpacing':
      return Math.abs(pixels(rendered) - figma) <= 0.02;
    case 'opacity':
      return Math.abs(Number(rendered) - figma) <= 0.01;
    case 'words':
      return rendered === figma;
    default:
      return Math.abs(pixels(rendered) - figma) <= 0.5;
  }
}

/** The oracle properties a renderer is measured on; the rest describe composition. */
export const MEASURED = [
  'background',
  'borderColor',
  'borderWidth',
  'borderTopWidth',
  'borderRightWidth',
  'borderBottomWidth',
  'borderLeftWidth',
  'radius',
  'radiusTopLeft',
  'radiusTopRight',
  'radiusBottomRight',
  'radiusBottomLeft',
  'shadow',
  'paddingTop',
  'paddingRight',
  'paddingBottom',
  'paddingLeft',
  'gap',
  'width',
  'height',
  'x',
  'y',
  'right',
  'bottom',
  'opacity',
  'color',
  'fontFamily',
  'fontWeight',
  'fontSize',
  'lineHeight',
  'letterSpacing',
  'textDecoration',
  'words',
];

/**
 * Every measured property of one layer, as failures and excused gaps.
 *
 * @param {object} expected the oracle's layer
 * @param {object} rendered the browser's computed values for that layer
 * @param {object[]} excused the variant's excused entries for this layer
 */
export function compareLayer(expected, rendered, excused = []) {
  const failures = [];
  const gaps = [];
  for (const property of MEASURED) {
    if (!(property in expected)) continue;
    const figma = expected[property];
    const got = rendered[property];
    // An excused entry is a gap even where nothing is drawn, so every excuse is seen reached.
    const excuse = excused.find((e) => e.property === property);
    if (!excuse && property === 'borderColor' && expected.borderWidth === 0)
      // A border colour on no border is not drawn.
      continue;
    if (excuse) {
      gaps.push({
        property,
        figma,
        rendered: got,
        finding: excuse.finding,
        decision: excuse.decision,
      });
      continue;
    }
    const want = rounded(property, figma, rendered);
    const drawn = CORNERS.has(property)
      ? rounded(property, pixels(got), rendered)
      : got;
    if (!matches(property, want, drawn))
      failures.push({ property, figma, rendered: got });
  }
  return { failures, gaps };
}

const CORNERS = new Set([
  'radius',
  'radiusTopLeft',
  'radiusTopRight',
  'radiusBottomRight',
  'radiusBottomLeft',
]);

/**
 * A corner no rounder than its box allows: no corner is drawn rounder than half the box's shorter
 * side, in CSS, in Flutter or in Figma, so a pill's 9999 and an ellipse's half its size draw the
 * same round. Compared as drawn, where the box was measured.
 */
function rounded(property, figma, rendered) {
  const [width, height] = [pixels(rendered.width), pixels(rendered.height)];
  if (!CORNERS.has(property) || !(width > 0) || !(height > 0)) return figma;
  return Math.min(figma, width / 2, height / 2);
}
